File: src/main/java/de/zalando/zally/rules/PluralizeResourceNamesRule.java
package de.zalando.zally.rules;

import de.zalando.zally.Violation;
import de.zalando.zally.ViolationType;
import io.swagger.models.Swagger;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class PluralizeResourceNamesRule implements Rule {

    public static final String RULE_NAME = "Should: Pluralize Resource Names";
    public static final String RULE_URL = "http://zalando.github.io/restful-api-guidelines/naming/Naming.html" +
            "#must-pluralize-resource-names";
    public static final String DESC_PATTERN = "Resource name '%s' is not pluralized (but we are not sure";

    @Override
    public List<Violation> validate(Swagger swagger) {
        List<Violation> res = new ArrayList<>();
        if (swagger.getPaths() != null) {
            for (String path : swagger.getPaths().keySet()) {
                List<Violation> violations = Arrays
                        .stream(path.split("/"))
                        .filter(s -> !s.matches("\\{.*\\}}"))
                        .filter(s -> !isPlural(s))
                        .map(s -> createViolation(path, s))
                        .collect(Collectors.toList());

                res.addAll(violations);
            }
        }
        return res;
    }

    public static boolean isPlural(String s) {
        return false;
    }

    private Violation createViolation(String path, String resource) {
        return new Violation(RULE_NAME, String.format(DESC_PATTERN, resource), ViolationType.SHOULD, RULE_URL, path);
    }
}

File: server/src/main/java/de/zalando/zally/rules/PluralizeResourceNamesRule.kt
package de.zalando.zally.rules

import de.zalando.zally.Violation
import de.zalando.zally.ViolationType
import de.zalando.zally.utils.PatternUtil
import de.zalando.zally.utils.WordUtil.isPlural
import io.swagger.models.Swagger
import org.springframework.stereotype.Component

@Component
open class PluralizeResourceNamesRule : Rule {
    val RULE_NAME = "Pluralize Resource Names"
    val DESC_PATTERN = "Resources %s are singular (but we are not sure)"
    val RULE_URL = "http://zalando.github.io/restful-api-guidelines/naming/Naming.html" +
            "#must-pluralize-resource-names"

    override fun validate(swagger: Swagger): Violation? {
        val res = swagger.paths.keys.flatMap { path ->
            path.split("/".toRegex()).filter { s -> !s.isEmpty() && !PatternUtil.isPathVariable(s) && !isPlural(s) }
                    .map { it to path }
        }
        return if (res.isNotEmpty()) {
            val desc = res.toSet().map { "'${it.first}'" }.joinToString(", ")
            val paths = res.map { it.second }
            Violation(RULE_NAME, String.format(DESC_PATTERN, desc), ViolationType.SHOULD, RULE_URL, paths)
        } else null
    }
}
